closed/src/java.base/share/classes/com/sun/crypto/provider/NativeChaCha20Cipher.java

```diff
@@ -88,19 +90,41 @@
     private ChaChaEngine engine;
 
     private static NativeCrypto nativeCrypto;
+    private static final Cleaner contextCleaner;
     private long context;
 
     private final ByteArrayOutputStream aadBuf;
 
     static {
         nativeCrypto = NativeCrypto.getNativeCrypto();
+        contextCleaner = CleanerFactory.cleaner();
+    }
+
+    private static final class ChaCha20CleanerRunnable implements Runnable {
+        private final long context;
+
+        public ChaCha20CleanerRunnable(long context) {
+            this.context = context;
+        }
+
+        @Override
+        public void run() {
+            /*
+             * Release the ChaCha20 context.
+             */
+            long ret = nativeCrypto.DestroyContext(context);
+            if (ret == -1) {
+                throw new ProviderException("Error in Native CipherBlockChaining");
+            }
+        }
     }
 
     /**
      * Default constructor.
      */
     protected NativeChaCha20Cipher() {
         context = nativeCrypto.CreateContext();
+        contextCleaner.register(this, new ChaCha20CleanerRunnable(this.context));
```

Comment:
I suggest this should be predicated on `context != -1`.

Reply:
done also added an exception if CreateContext() returns an error